### PR TITLE
Add cis3-nft-sponsored-txs

### LINKED ISSUE #222: Create a Cis3-NFT contract that works with gasless transactions.
**Task description**

The contract will need a registry for mapping `publicKey` to the `accountAddress` before a gasless transaction can be processed in the smart contract.

### PR DESCRIPTION
**Pull request by DOBEN**
## Purpose

closes #222

Next step: Move parts of this example to a CIS3 library.

## Changes

New functions:
- `registerPublicKey` (only the owner or account itself can register a public key)
- `publicKeyOf` (query the public keys of a list of given accounts)
- `permit` (sponsored transaction counterpart to the regular `updateOperator/transfer` function)

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:747]: We don't really use the `_` prefix for internal functions. It is internal since it doesn't have a `receive` attribute, or a `pub` keyword in front 😊

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:856]:
```suggestion
/// - Any of the receive hook function calls reject.
```

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:929]: `if is_valid` is probably more idiomatic 😊

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:1050]: If `params.len > 1` then it is cheaper to keep the `nonce` in a variable outside the loop and only write it to the state once the for loop has finished.

**Review by Bargsteen — APPROVED** [review]
Nice work! I just had a few minor findings.

Regarding the batched registration of public keys: I think we should add that. It will just be a few extra bytes (dependent on the `size_length` we choose), but will reduce the number of transactions if someone wants to batch these.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:51]: I removed that sentence about the trust assumption for the third-party service.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:856]: Because of the *any/anyone* (singular), I feel it should be `rejects`.

**Any** of the receive hook function calls rejects

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:1050]: The `param.owner` can be different in every loop, so also the nonce can be different in every loop and that is why the nonce has to be inside of the loop and updated for every iteration in the for loop.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:597]: I added it and also a test for this view function.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs]: Updated the sentence to:

```
Once an account has a public key registered, the mapping between the public key and the account can not be updated anymore. 
```

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:1070]: Ah, right. That was a blunder on my part ;)

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:601]: Great 😊

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:639]: If you use a set (fx `BTreeSet`), then you can avoid duplicates automatically.
We also support sets in the schema, and our tools (at least concordium-client) will check whether it contains any duplicates before making a transaction.

**Review by Bargsteen — APPROVED** [review]
LGTM if you use a set ;)

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (limemloh) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:860]: I would limit the entry point to only take one signed message instead of a list.

Also, the signed message should contain a list of transfers, like in CIS-2.
Currently, you cannot sign multiple transfer in one message, and signing multiple messages allows the account sponsoring the transaction to split it up, which might not be ideal.

**Review by limemloh — CHANGES_REQUESTED** [review]

**Review by abizjak — CHANGES_REQUESTED** [review]
I haven't gone through all the details, because there seemed to have been a misunderstanding.

The idea was to have a standard, CIS3, specifying how to separate the sender from the person intending to do the action. That would be specifying essentially a format like the PermitTransferParams which would have

- signature
- signer
- nonce
- timestamp
- parameters of UpdatePayload like a normal transaction has (minus the amount I suppose since that should be 0 always, otherwise there probably isn't much point)

Then this example would be an extension of a standard token that would add this extra endpoint and dispatch to the normal endpoint of the contract.

Does that make sense?

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:560]: I kept it for now since it is the same pattern as the `balanceOf` and `operatorOf` functions to keep it consistent.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:1059]: yes, I extended the getter with the current nonce.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs]: Support transfers from contracts would mean that the smart contract needs to `sign` a transaction but we don't have a standard yet for that. For now, it seems that the `permit` function can be used by only `AccountAddresses` since they have public keys and can sign.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (abizjak) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:279]: So I think there is a bit of an annoying issue here.

The serialization of `PermitPayload` will have a one byte tag, and then the transfer params or update operator params.

This means that in the CIS3 standard we would have to specify that the "payload" should have this extra byte tag or something like it instead of just the payload of the entrypoint. That's not necessarily bad, but needs to be clarified.

What exactly do we want here? Do we want to allow the permit to add extra data in addition to the parameters of the entrypoints?

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (abizjak) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:288]: I think we should support the same type of signature, which is a multisig like we do for transactions.

That will make it simpler to later use accounts. That signature is essnetially just a map
```
BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>
```

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (abizjak) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:817]:
```suggestion
        .ok_or(CustomContractError::NoPublicKey)?;
```

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (abizjak) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:819]:
```suggestion
    let public_key = entry.as_ref().0;
    let nonce = entry.as_ref().1;
```

Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:819]: Changed it to 

```
    let public_key = entry.0;
    let nonce = entry.1;
```

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (limemloh) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:279]: Could we make the payload generic and leave it to the implementation?

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs]: I implemented the following: The CIS3 standard supports multi-sig accounts but because the `public_key_registry` in this contract maps one public key to one account, only one signature has to be provided for this contract. This signature has to be at the key 0 in both maps below:
```
  signature: BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>
 ```

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (Bargsteen) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:279]: If we allow any type of payload, then we can completely remove the link to CIS-2, right?
Then CIS-3 can become a universal standard for sponsored transactions, which I think is useful.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (abizjak) [inline examples/cis3-nft-sponsored-txs/src/lib.rs:902]:
```suggestion
    // Update the nonce.
    let mut entry = host.state_mut().public_key_registry.entry(param.signer).occupied_or(CustomContractError::NoPublicKey)?;
    entry.1 += 1; // bump nonce
    // Get the public key and the current nonce.
    let public_key = entry.0;
    let nonce = entry.1;
    drop(entry);
```

Comment (abizjak) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:902]: To avoid a double lookup and make the code a bit simpler.

**Review by abizjak — APPROVED** [review]
You need to merge with main.

But otherwise this looks reasonable.

Maybe there will be some changes needed when developing the frontend and writing the standard, but for now this looks OK to me.

**Review by limemloh — APPROVED** [review]
Looks good!
Only some minor comments/considerations

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:873]: There is a `wrongEntryPoint` error in the smart contract already which is thrown when the specified `entry_point` is accidentally swapped around (e.g. somebody wants to call the `transfer` function but provides the `operatorInputParamter`).

https://github.com/Concordium/concordium-rust-smart-contracts/blob/d00aa22d7f6e17111c5af6f6ee211bc947a8537e/examples/cis3-nft-sponsored-txs/src/lib.rs#L326

The embedded schema in the smart contract will specify which input parameters are supported for the `permit` function, so I hope users will make use of the schema and not come up with other inputParameters/functionEntryPoints that they think could be supported. In general, it will be difficult to figure out if this input Parameter deserialization fails due to a small mistake in the inputParameter provided or if the user completely misunderstood it and thought that additional entryPoints/inputParameters are supported. I suggest keeping it for now but I am open to suggestions if there are improvements we can do done the line.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (DOBEN) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:942]: I will bring this up for discussion tomorrow and implement in a new MR if we want to do this change.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (limemloh) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:874]: While it is helpful, we should not (IMO) rely on the schema to provide information like this.
We could have an endpoint for endpoints supporting sponsored transactions, but a good error code is still useful.

File: examples/cis3-nft-sponsored-txs/src/lib.rs
Comment (abizjak) [reply examples/cis3-nft-sponsored-txs/src/lib.rs:942]: To finalize this, I think the agreement at the meeting was to do this for consistency, not that it would likely have any specific utility.

**Pull request merged**
